### DISCUSSION TITLE
How do you keep the SSE connection open longer than 5 mins?

### DISCUSSION THREAD

**Discussion by AFatalErrror** [pocketbase/pocketbase#3054] [Q&A]
I am trying to keep the SSE connection "/api/realtime" open indefinitely. 

Thank you!

**Comment by ganigeorgiev** [comment]
The 5 mins idle timeout is "hard coded" and not configurable at the moment (I'll consider making it eventually configurable but that will be handled at some later stage with the planned per request timeout handling).

I'm not sure what is your use case but in general I'd advice against keeping the SSE connection open indefinitely as it could have unintended side-effects in case of invalid http client implementations or forgotten/"zombie" connections.

**Reply by AFatalErrror** [reply]
is there a way to ping that connection to let pocketbase know its still alive?

**Reply by ganigeorgiev** [reply]
I haven't tested it, but I guess one hacky workaround could be to iterate over the subscribed clients and push a dummy message periodically. If you use the JS app hooks, this could look something like:

```js
// send a dummy ping message every 3 minutes
cronAdd("ping", "*/3 * * * *", () => {
    const msg = new DynamicModel({ name: "ping" });
    const clients = $app.subscriptionsBroker().clients()
    for (const clientId in clients) {
        clients[clientId].send(msg)
    }
})
```

**Reply by ganigeorgiev** [reply]
_I'll consider with the next release to add the idle timer to the `core.RealtimeConnectEvent` struct so that you'll be able to do something like:_

```js
onRealtimeConnectRequest((e) => {
    // e.idleDuration = duration
    e.idleTimer.reset(duration)
})
```

**Reply by AFatalErrror** [reply]
awesome, thanks alot!

**Reply by ganigeorgiev** [reply]
Just for info, I've added a new `RealtimeConnectEvent.IdleTimeout` field in `master` that should allow changing the timeout per client basis. The change will be part of the next minor v0.17.2 release sometime later this week.

_Note that in order to avoid having and maintain 2 different APIs, the timer type is `time.Duration`, aka. in nanosec:_

```go
// Go
app.OnRealtimeConnectRequest().Add(func(e *core.RealtimeConnectEvent) error {
    e.IdleTimeout = 10 * time.Minute
})
```

```js
// JS
onRealtimeConnectRequest((e) => {
    e.idleTimeout = 10 * 60000000000 // 10 min
})
```
_(I'll consider adding eventually similar `time.*` helpers as those in Go)_

**Reply by calumk** [reply]
Can i ask for some clarity on the behaviour here:

https://pocketbase.io/docs/api-realtime/#connect

> If the connected client doesn't receive any new messages for 5 minutes, the server will send a disconnect signal (this is to prevent forgotten/leaked connections). The connection will be automatically reestablished if the client is still active (eg. the browser tab is still open).

So pocketbase will attempt to keep the conenction live, it will close it and re-open it, essentially allowing an 'indefinite' connection

Is there a way (from the client) to check if a subscription is still active?

Use case : 
 * Client connects to subscription
 * Client experiences some network issue, or maybe even pocketbase is closed / crashed?
 * How can the client know the connection is broken ?

**Reply by ganigeorgiev** [reply]
> Is there a way (from the client) to check if a subscription is still active?

If the connection with the server is broken the SDKs will try to perform autoreconnect behind the scene indefinitely on various intervals (200ms, 300ms, 500ms, etc. it will depend on the SSE implementation). Users usually shouldn't do anything in addition.

The JS SDK has a helper `pb.realtime.isConnected` getter that could be used to determine whether an SSE connection was established.

The Dart SDK currently doesn't have such helper exposed but I guess you can check if `pb.realtime.clientId` is non-empty.

If you want to know when the SSE connection was established (initially or on reconnect), you can subscribe to the `PB_CONNECT` event, aka. `pb.realtime.subscribe("PB_CONNECT", ...)`.

Currently we don't have direct event listener exposed in the SDKs that will get triggered on closed/restarted connection (_I've added it in my todo to consider it but for now remains a low priority_).
